src/quad_expr.jl

```diff
@@ -232,4 +232,20 @@ function Base.copy(q::GenericQuadExpr, new_model::Model)
                     copy(q.qcoeffs), copy(q.aff, new_model))
 end
 
-# TODO: result_value for QuadExpr
+# Requires that foo(::GenericAffExpr) is defined.
+function value(ex::GenericQuadExpr{CoefType, VarType},
+               foo::Function) where {CoefType, VarType}
+    # The return type of appling map to ::VarType.
+    MapVarType = Base.promote_op(foo, VarType)
+    # Later, we're going to multiply two MapVarType together
+    MapVarType2 = Base.promote_op(*, MapVarType, MapVarType)
+    # We're also going to multiply a constant with ::MapVarType2
+    RetType = Base.promote_op(*, CoefType, MapVarType2)
```

Comment:
I was once told that it is not safe to call `promote_op` on top of `promote_op`, see https://github.com/JuliaLang/julia/pull/18218#issuecomment-242331704.
So maybe something like `RetType = Base.promote_op((c, v) -> c * foo(v) * foo(v), CoefType, VarType)` is better.

Reply:
I just followed
https://github.com/JuliaOpt/JuMP.jl/blob/6daac7ae3105d1c2ddcd44a1324acd3769fcef64/src/aff_expr.jl#L105-L113

Reply:
I never understood the argument so I am not surprised to see that I am not following it :-P
In fact, it appears that using `promote_op` is just bad: https://github.com/JuliaLang/julia/pull/26344#issuecomment-371613750
So the best way is just to play with `typeof`, `zero`, `one` and `oneunit`. To choose whether to use `one` or `oneunit` you need to think what would happen if the type was frim Unitful. For choosing between `zero` and `one`, you have some freedom but you need to avoid dividing by zero.
Maybe the best way here is doing `typeof(oneunit(CoefType) * oneunit(MapVarType) * oneunit(MapVarType))`. We are kind of forced to use `promote_op` to get `MapVarType` since `one` ... would on `JuMP.VariableRef` would not really do what we want.